doc/dms.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DMS" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBdms\fR \- Grow a scale\-free random graph with tunable exponent
.
.SH "SYNOPSIS"
\fBdms\fR \fIN\fR \fIm\fR \fIn0\fR \fIa\fR
.
.SH "DESCRIPTION"
\fBdms\fR grows an undirected random scale\-free graph with \fIN\fR nodes using the modified linear preferential attachment model proposed by Dorogovtsev, Mendes and Samukhin\. The initial network is a clique of \fIn0\fR nodes, and each new node creates \fIm\fR new edges\. The resulting graph will have a scale\-free degree distribution, whose exponent converges to \fBgamma=3\.0 + a/m\fR for large \fIN\fR\.
.
.SH "PARAMETERS"
.
.TP
\fIN\fR
Number of nodes of the final graph\.
.
.TP
\fIm\fR
Number of edges created by each new node\.
.
.TP
\fIn0\fR
Number of nodes in the initial (seed) graph\.
.
.TP
\fIa\fR
This parameter sets the exponent of the degree distribution (\fBgamma = 3\.0 + a/m\fR)\. \fIa\fR must be larger than \fI\-m\fR\.
.
.SH "OUTPUT"
\fBdms\fR prints on STDOUT the edge list of the final graph\.
.
.SH "EXAMPLES"
Let us assume that we want to create a scale\-free network with \fIN=10000\fR nodes, with average degree equal to 8, whose degree distribution has exponent
.
.IP "" 4
.
.nf

    gamma = 2\.5
.
.fi
.
.IP "" 0
.
.P
Since \fBdms\fR produces graphs with scale\-free degree sequences with an exponent \fBgamma = 3\.0 + a/m\fR, the command:
.
.IP "" 4
.
.nf

    $ dms 10000 4 4 \-2\.0 > dms_10000_4_4_\-2\.0\.txt
.
.fi
.
.IP "" 0
.
.P
will produce the desired network\. In fact, the average degree of the graph will be:
.
.IP "" 4
.
.nf

    <k> = 2m = 8
.
.fi
.
.IP "" 0
.
.P
and the exponent of the power\-law degree distribution will be:
.
.IP "" 4
.
.nf

    gamma = 3\.0 + a/m = 3\.0 \-0\.5 = 2\.5
.
.fi
.
.IP "" 0
.
.P
The following command:
.
.IP "" 4
.
.nf

    $ dms 10000 3 5 0 > dms_10000_3_5_0\.txt
.
.fi
.
.IP "" 0
.
.P
creates a scale\-free graph with \fIN=10000\fR nodes, where each new node creates \fIm=3\fR new edges and the initial seed network is a ring of \fIn0=5\fR nodes\. The degree distribution of the final graph will have exponent equal to \fBgamma = 3\.0 + a/m = 3\.0\fR\. In this case, \fBdms\fR produces a Barabasi\-Albert graph (see ba(1) for details)\. The edge list of the graph is saved in the file \fBdms_10000_3_5_0\.txt\fR (thanks to the redirection operator \fB>\fR)\.
.
.SH "SEE ALSO"
ba(1), bb_fitness(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
S\. N\. Dorogovtsev, J\. F\. F\. Mendes, A\. N\. Samukhin\. "Structure of Growing Networks with Preferential Linking"\. Phys\. Rev\. Lett\. 85 (2000), 4633\-4636\.
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 6, Cambridge University Press (2017)
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Appendix 13, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/dms/dms.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 * This program implements the Dorogovtsev-Samukhin-Mendes preferential
 * attachment, where the attachment probability is:
 *
 * \Pi_{i->j} \propto k_j + a
 *
 * Here a > -m is a tunable parameter. The resulting network has a
 * powerlaw degree distribution with exponent:
 *
 * \gamma = 3 + a/m
 *
 *  References:
 *  
 * [1] S. N. Dorogovtsev, J. F. F. Mendes, A. N. Samukhin. "Structure
 *     of Growing Networks with Preferential Linking".
 *     Phys. Rev. Lett. 85 (2000), 4633-4636. 
 *
 */


#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "cum_distr.h"


void usage(char *argv[]){
  printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-    dms    -*-                       **\n"
         "**                                                                **\n"
         "**   Grow a scale-free network of 'N' nodes using the modified    **\n"
         "**   linear preferential attachment model proposed by             **\n"
         "**   Dorogovtsev-Mendes-Samukhin.                                 **\n"
         "**                                                                **\n"
         "**   The initial network is a clique of 'n0' nodes, and each new  **\n"
         "**   node creates 'm' edges. The attachment probability is of     **\n"
         "**   the form:                                                    **\n"
         "**                                                                **\n"
         "**            P(i->j) ~ k_j + a                                   **\n"
         "**                                                                **\n"
         "**   where a > -m is the fourth parameter. The resulting          **\n"
         "**   network will have a power-law degree distribution with       **\n"
         "**   exponent                                                     **\n"
         "**                                                                **\n"
         "**       gamma = 3 + a/m                                          **\n"
         "**                                                                **\n"
         "**   The program prints on STDOUT the edge-list of the final      **\n"
         "**   graph.                                                       **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " (c) Vincenzo Nicosia 2009-2017 (v.nicosia@qmul.ac.uk)\n\n"
         "********************************************************************\n\n"
         );
  printf("Usage: %s <N> <m> <n0> <a>\n", argv[0]);

}



int init_network(unsigned int *I, unsigned int *J, int n0, 
                 double a, cum_distr_t *d){
  
  unsigned int n, i,  S_num;
  
  S_num = 0;
  for(n=0; n<n0; n++){
    for(i=n+1; i<n0; i++){
      I[S_num] = n;
      J[S_num] = i % n0;
      S_num += 1;
    }
    cum_distr_add(d, n, n0+a);
  }
  return S_num;
}

int already_neighbour(unsigned int *J, int S_num, int j, int dest){

   int i;

   for(i=S_num; i< S_num + j; i ++){
     if (J[i] == dest)
       return 1;
   }
   return 0;
}



int dms(unsigned int *I, unsigned int *J, unsigned int N, 
         unsigned int m, unsigned int n0,  double a){
  
  cum_distr_t *d = NULL;
  unsigned int n, j, dest, S_num;
  
  d = cum_distr_init(N * m);
  
  S_num = init_network(I, J, n0, a, d);
  
  
  n = n0;
  while (n<N){
    for(j=0; j<m; j++){
      I[S_num+j] = n;
      dest = cum_distr_sample(d);
      while(already_neighbour(J, S_num, j, dest)){
        dest = cum_distr_sample(d);
      }
      J[S_num + j] = dest;
    }
    cum_distr_add(d, n, m + a);
    for (j=0; j<m; j++){
      cum_distr_add(d, J[S_num + j], 1);
    }
    S_num += m;
    n += 1;
  }
  cum_distr_destroy(d);
  return S_num;
}



int main(int argc, char *argv[]){
  
  int N, m, n0, K, i;
  unsigned int *I, *J;
  double a;
  
  if (argc < 5){
    usage(argv);
    exit(1);
  }
  
  N = atoi(argv[1]);
  m = atoi(argv[2]);
  n0 = atoi(argv[3]);
  a = atof(argv[4]);

  srand(time(NULL));

  if (N < 1){
    fprintf(stderr, "N must be positive\n");
    exit(1);
  }
  if(m > n0){
    fprintf(stderr, "n0 cannot be smaller than m\n");
    exit(1);

  }
  if (n0<1){
    fprintf(stderr, "n0 must be positive\n");
    exit(1);
  }
  
  if (m < 1){
    fprintf(stderr, "m must be positive\n");
    exit(1);
  }

  if (a < -m){
    fprintf(stderr, "a must be larger than -m\n");
    exit(1);
  }
  
  
  I = malloc(N * m * sizeof(unsigned int));
  J = malloc(N * m * sizeof(unsigned int));
  
  K = dms(I, J, N, m, n0, a);
  
  for(i=0; i<K; i++){
    printf("%d %d\n", J[i], I[i]);
  }
  free(I);
  free(J);
}
